New Card: Shopping List

iantrich opened the pull request:
Following features:
- Add item
- Edit item
- Complete item
- Clear completed items

![shopping-list](https://user-images.githubusercontent.com/1287159/47947939-a962a180-def4-11e8-9901-f1f224e5f110.gif)

iantrich committed:
New Card: Shopping List

Following features:
- Add item
- Edit item
- Complete item
- Clear items

iantrich committed:
Address Travis complaint

zsarnett commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,261 @@
+    }
+  }
+
+  private _saveEdit(ev): void {
```
Check for hass. You check all other places but here and in complete Item

zsarnett commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,261 @@
+class HuiShoppingListCard extends hassLocalizeLitMixin(LitElement)
+  implements LovelaceCard {
+  public hass?: HomeAssistant;
+  protected _config?: Config;
```
Should be private,

iantrich committed:
Addressed review comments

iantrich committed:
Update translation variable name

iantrich committed:
Line up input row text

zsarnett commented:
What about something like this
![example](https://i.gyazo.com/e8a739be9b083d93d8e0f25724badf3b.png)

Adding the second headar (smaller) with the icon next to that

balloob commented on gallery/src/data/provide_hass.js:
```diff
@@ -29,6 +29,12 @@ export default (elements, { initialStates = {} } = {}) => {
     resources: demoResources,
     states: initialStates,
     themes: {},
+    connection: {
```
connection: Connection,

balloob commented on src/data/shopping-list.ts:
```diff
@@ -0,0 +1,39 @@
+    name,
+  });
+
+export const clearItems = (hass: HomeAssistant): Promise<void> =>
```
`clearItems` -> `clearCompleted`

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,273 @@
+  public connectedCallback(): void {
+    super.connectedCallback();
+
+    if (this.hass) {
```
So since you have this check, it is important that in case we get connected to the DOM and then get `hass`, we recover from that?

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,273 @@
+    super.connectedCallback();
+
+    if (this.hass) {
+      this.hass.connection
```
Make this `this._unsubEvents = this.hass.connection.subscribeEvents(…)` and then in `disconnected` do this:

```ts
if (this._unsubEvents) {
  this._unsubEvents.then(unsub => unsub());
}
```

That way, if the shopping card disconnects before the subscription is in place, we still unsubscribe.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,273 @@
+      <ha-card .header="${this._config.title}">
+        <ha-icon
+          id="clear"
+          icon="hass:notification-clear-all"
```
`.icon` ?

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,273 @@
+      ${repeat(
+        this._items!,
+        (item) =>
+          !item.complete
```
Just filter before passing it to the render function: `this._items.filter(item => !item.complete)`

Also add an identification function to repeat as 2nd arg: `item => item.id`

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,273 @@
+      ${repeat(
+        this._items!,
+        (item) =>
+          item.complete
```
use filter and add identify func.

balloob commented on src/panels/lovelace/cards/hui-shopping-list-card.ts:
```diff
@@ -0,0 +1,273 @@
+  }
+
+  private _addItem(ev): void {
+    if (this.shadowRoot!.querySelector("#addBox")) {
```
Don't ever write IFs that span whole or most of a function. Convert those to a guard clause:

```ts
const addBox = this.shadowRoot!.querySelector("#addBox");

if (!addBox) { return; }

…
```

balloob commented on src/translations/en.json:
```diff
@@ -394,6 +394,10 @@
       "script": {
         "execute": "Execute"
       },
+      "shopping-list": {
```
We should have a Lovelace section in the translation files.

iantrich committed:
Taking MVP to heart

Addressed review comments and scaled this back to just get a simple shopping list card out there and we can discuss/debate how best to add the additional pieces with smaller PRs

zsarnett commented:
I don't think we should limit this card to `shopping` List. I think List Card may be a better name for all around use

iantrich commented:
> I don't think we should limit this card to `shopping` List. I think List Card may be a better name for all around use

The component is 'Shopping List'. And I already made a list-card :)
https://github.com/custom-cards/list-card

zsarnett commented:
> The component is 'Shopping List'. And I already made a list-card :)
> https://github.com/custom-cards/list-card

Ahh Gotcha. Fair enough.

iantrich commented:
> > The component is 'Shopping List'. And I already made a list-card :)
> > https://github.com/custom-cards/list-card
> 
> Ahh Gotcha. Fair enough.

In the future I plan to update the shopping list component to support multiple lists and would probably propose a renaming, but for MVP, I think this fits

balloob committed:
Remove calling connected in set hass

balloob commented:
I've removed the calling of connectedCallback inside the `hass` setter as it had two unintended consequences:
 - we ended up calling `connectedCallback` of all super classes 
 - we ended up subscribing before connected, and then when we got connected, we would subscribe a second time.

If I missed something or broke something, we can fix it in another PR. Yay for dev branch.
